batch.py

```diff
@@ -0,0 +1,356 @@
+
+    Input:
+    gene_file: path to a tsv file with information on
+    the chrom, start and end location of genes (rows)
+
+    Output:
+    For retained variants, that are: 1) not ref-only, 2) biallelic,
+    3) meet QC filters, 4) rare (MAF<5%), 5) in promoter regions and
+    6) within 50kb up or down-stream of the gene body (or in the gene body itself)
+    returns:
+    - Path to plink prefix for genotype files (plink format: .bed, .bim, .fam)
+    - Path to hail table with variant (rows) stats for downstream analyses
+    """
+    # read hail matrix table object (WGS data)
+    init_batch()
+    mt = hl.read_matrix_table(mt_path)
+
+    # get relevant chromosome
+    gene_df = pd.read_csv(AnyPath(dataset_path(gene_file)), sep="\t", index_col=0)
+    chrom = gene_df[gene_df["gene_name"] == gene_name]["chr"]
+    # subset to chromosome
+    mt = mt.filter_rows(mt.locus.contig == ("chr" + chrom))
+
+    # densify
+    mt = hl.experimental.densify(mt)  # never know when i should do this step
+
+    # subset to window
+    # get gene body position (start and end) and build interval
+    interval_start = int(gene_df[gene_df["gene_name"] == gene_name]["start"]) - int(
+        window_size
+    )
+    interval_end = int(gene_df[gene_df["gene_name"] == gene_name]["end"]) + int(
+        window_size
+    )
+    # clip to chromosome boundaries
+    left_boundary = max(1, interval_start)
+    right_boundary = min(
+        interval_end, hl.get_reference("GRCh38").lengths[f"chr{chrom}"]
+    )
+    # get gene-specific genomic interval
+    gene_interval = f"chr{chrom}:{left_boundary}-{right_boundary}"
+    logging.info(f"Interval considered: {gene_interval}")  # 'chr22:23219960-23348287'
+
+    # include variants up to {window size} up- and downstream
+    mt = hl.filter_intervals(
+        mt, [hl.parse_locus_interval(gene_interval, reference_genome="GRCh38")]
+    )
+    logging.info(f"Number of variants within interval: {mt.count()[0]}")
+    logging.info(f"Number of variants in window: {mt.count()[0]}")
+
+    # filter out low quality variants and consider biallelic variants only (no multi-allelic, no ref-only)
+    mt = mt.filter_rows(
+        (hl.len(hl.or_else(mt.filters, hl.empty_set(hl.tstr))) == 0)  # QC
+        & (hl.len(mt.alleles) == 2)  # remove hom-ref
+        & (mt.n_unsplit_alleles == 2)  # biallelic
+        & (hl.is_snp(mt.alleles[0], mt.alleles[1]))  # SNVs
+    )
+
+    # annotate using VEP
+    vep_ht = hl.read_table(ht_path)
+    mt = mt.annotate_rows(vep=vep_ht[mt.row_key].vep)
+    mt_path = output_path("densified_qced_snps_only_vep_annotated.mt", "tmp")
+    mt = mt.checkpoint(
+        mt_path, overwrite=True
+    )  # add checkpoint to avoid repeat evaluation
+    logging.info(f"Number of QC-passing, biallelic SNPs: {mt.count()[0]}")
+
+    # filter variants found to be in promoter regions
+    mt = mt.filter_rows(
+        mt.vep.regulatory_feature_consequences["biotype"][0] == "promoter"
+    )
```

Comment:
is this always the first/only annotation?

Reply:
it is the only one i am using for this first run, but hope to add more in the future!

Reply:
Cool, I was just wondering if you can always rely on 'promoter' being the first in this list. It could be safer to check that 'promoter' is just somewhere in the list

```suggestion
    mt = mt.filter_rows(
        mt.vep.regulatory_feature_consequences["biotype"].contains("promoter")
    )
```

Reply:
yep good catch, thanks!